xarray/core/variable.py

```diff
@@ -94,7 +94,8 @@ def as_variable(obj, name=None):
                             '{}'.format(obj))
     elif utils.is_scalar(obj):
         obj = Variable([], obj)
-    elif getattr(obj, 'name', None) is not None:
+    elif (isinstance(obj, (pd.Index, IndexVariable)) and
+          getattr(obj, 'name', None) is not None):
```

Comment:
This can just be `obj.name is not None` (since these objects are guaranteed to have `name` attributes)